x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/index/IndexResolver.java

```diff
@@ -242,103 +243,82 @@ private void filterResults(String javaRegex, GetAliasesResponse aliases, GetInde
     public void resolveAsMergedMapping(String indexWildcard, String javaRegex, ActionListener<IndexResolution> listener) {
         FieldCapabilitiesRequest fieldRequest = createFieldCapsRequest(indexWildcard);
         client.fieldCaps(fieldRequest,
-                ActionListener.wrap(response -> listener.onResponse(mergedMapping(indexWildcard, response.get())), listener::onFailure));
+                ActionListener.wrap(
+                        response -> listener.onResponse(mergedMappings(indexWildcard, response.getIndices(), response.get())),
+                        listener::onFailure));
     }
 
-    static IndexResolution mergedMapping(String indexPattern, Map<String, Map<String, FieldCapabilities>> fieldCaps) {
+    static IndexResolution mergedMappings(String indexPattern, String[] indexNames, Map<String, Map<String, FieldCapabilities>> fieldCaps) {
         if (fieldCaps == null || fieldCaps.isEmpty()) {
             return IndexResolution.notFound(indexPattern);
         }
 
-        StringBuilder errorMessage = new StringBuilder();
+        // merge all indices onto the same one
+        List<EsIndex> indices = buildIndices(indexNames, null, fieldCaps, i -> indexPattern, (n, types) -> {
+            StringBuilder errorMessage = new StringBuilder();
 
-        NavigableSet<Entry<String, Map<String, FieldCapabilities>>> sortedFields = new TreeSet<>(
-                // for some reason .reversed doesn't work (prolly due to inference)
-                Collections.reverseOrder(Comparator.comparing(Entry::getKey)));
-        sortedFields.addAll(fieldCaps.entrySet());
-
-        Map<String, EsField> hierarchicalMapping = new TreeMap<>();
-        Map<String, EsField> flattedMapping = new LinkedHashMap<>();
-        
-        // sort keys descending in order to easily detect multi-fields (a.b.c multi-field of a.b)
-        // without sorting, they can still be detected however without the emptyMap optimization
-        // (fields without multi-fields have no children)
-        for (Entry<String, Map<String, FieldCapabilities>> entry : sortedFields) {
+            boolean hasUnmapped = types.containsKey(UNMAPPED);
 
-            InvalidMappedField invalidField = null;
-            FieldCapabilities fieldCap = null;
-            errorMessage.setLength(0);
+            if (types.size() > (hasUnmapped ? 2 : 1)) {
+                // build the error message
+                // and create a MultiTypeField
 
-            String name = entry.getKey();
+                for (Entry<String, FieldCapabilities> type : types.entrySet()) {
+                    // skip unmapped
+                    if (UNMAPPED.equals(type.getKey())) {
```

Comment:
Personal preference: I'd inverse the check and surround the whole block with this if, rather than use `continue`

Reply:
I'm not a fan of continue but I prefer it instead of indenting a long method already wrapped...

Reply:
Yep, I understand, and I'm not insisting, I didn't comment at all here: https://github.com/elastic/elasticsearch/pull/41377/files/2e890c11f3283d32c6325f7d9ebb7d14b1972ec5#diff-e9f288c0280d1beafe4dc4cf86c12c7bR433 where the block is even larger as inverting the if makes it much less readable.